### PR TITLE
3410 http agent binary server

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of #3410.

Adds the HTTPAgentBinaryServer

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [x] Do all end-to-end tests pass?
* [ ] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/infection_monkey/exploit/http_agent_binary_server.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        thread_name = f"HTTPAgentBinaryServer-{insecure_generate_random_string(n=8)}"
+        return create_daemon_thread(
+            target=server.serve_forever,
+            name=thread_name,
+            args=(self._poll_interval,),
+        )
+
+    def deregister(self, request_id: RequestID) -> None:
+        """
+        Unregister an Agent binary from being served
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/http_agent_binary_server.py:141]:
```suggestion
        Deregister an Agent binary from being served
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

**Comment by mssalvatore** [comment]
We'll need to fix the commits that start with `TEMP`.

File: monkey/infection_monkey/exploit/agent_binary_request.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+RequestID: TypeAlias = UUID
+
+
+class RequestType(Enum):
+    AGENT_BINARY = "agent_binary"
+    DROPPER_SCRIPT = "dropper_script"
+
+
+@dataclass(frozen=True)
+class AgentBinaryRequest:
```
[the remark this answers is not in the document]
Comment (cakekoa) [reply monkey/infection_monkey/exploit/agent_binary_request.py:20]: Hmm... Download... Chit? Receipt? Ticket? Token? Registration?

[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/infection_monkey/exploit/agent_binary_request.py:19]: Record? Resource?

This one is hard.

File: monkey/infection_monkey/exploit/agent_binary_request.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    id: RequestID
+    type: RequestType
+    operating_system: OperatingSystem
+    destination_path: Optional[PurePath]
+    args: Sequence[str]
+    download_url: str
+    bytes_downloaded: Event
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/agent_binary_request.py:26]: Suggestion:
```suggestion
    download_completed: Event
```

File: monkey/infection_monkey/exploit/agent_binary_request.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/infection_monkey/exploit/agent_binary_request.py:19]: After discussion, we've decided to produce two dataclasses. The one that gets returned to the user will be `AgentBinaryDownloadTicket` and the one stored internally in the handler is `AgentBinaryDownloadReservation`.

**Comment by cakekoa** [comment]
This works, but comes with a caveat: Since the `HTTPAgentBinaryServer` is run in a separate process and plugins are loaded dynamically, the `HTTPAgentBinaryServer` will have trouble with an `AgentBinaryTransform` that is defined by the plugin, since it won't know how to look it up in order to serialize it

**Review by shreyamalviya — COMMENTED** [review]
Nice job!

File: monkey/infection_monkey/exploit/http_agent_binary_server.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        agent_binary_transform: AgentBinaryTransform = use_agent_binary,
+    ) -> AgentBinaryDownloadTicket:
+        """
+        Register to download an Agent binary
+
+        If the server is not running, it will be started.
+
+        :param operating_system: The operating system for the Agent binary to serve
+        :param requestor_ip: The IP address of the client that will download the Agent binary
+        :param agent_binary_transform: A function that transforms the Agent binary before serving.
```
Comment (shreyamalviya) [inline monkey/infection_monkey/exploit/http_agent_binary_server.py:75]: I have a vague idea right now that this has something to do with the dropper script, but can we make this more descriptive? Someone reading this for the first time can't make much of it.

**Review by ilija-lazoroski — APPROVED** [review]

**Pull request merged**